statsmodels/miscmodels/ordinal_model.py

```diff
@@ -0,0 +1,248 @@
+                                  "categories. ordered == True preferred.",
+                                  Warning)
+                endog_name = endog.name
+                threshold_name = [str(x) + '/' + str(y)
+                                  for x, y in zip(endog.values.categories[:-1],
+                                                  endog.values.categories[1:])]
+                names['yname'] = endog_name
+                names['xname'] = names['xname'] + threshold_name
+                endog = np.asarray(endog.values.codes)
+            else:
+                msg = (
+                    "If the endog is a pandas.Serie "
+                    "it must be of categoricalDtype."
+                )
+                raise ValueError(msg)
+
+        return names, endog, exog
+
+    def cdf(self, x):
+        """cdf evaluated at x
+        """
+        return self.distr.cdf(x)
+
+    def prob(self, low, upp):
+        """interval probability
+        """
+        return np.maximum(self.cdf(upp) - self.cdf(low), 0)
+
+    def transform_threshold_params(self, params):
+        """transformation of the parameters in the optimization
+
+        Parameters
+        ----------
+        params : nd_array
+            contains (exog_coef, transformed_thresholds) where exog_coef are
+            the coefficient for the explanatory variables in the linear term,
+            transformed threshold or cutoff points. The first, lowest threshold
+            is unchanged, all other thresholds are in terms of exponentiated
+            increments
+
+        Returns
+        -------
+        thresh : nd_array
+            thresh are the thresholds or cutoff constants for the intervals.
+
+        """
+        th_params = params[-(self.k_levels - 1):]
+        thresh = np.concatenate((th_params[:1],
+                                 np.exp(th_params[1:]))).cumsum()
+        thresh = np.concatenate(([-np.inf], thresh, [np.inf]))
+        return thresh
+
+    def transform_reverse_threshold_params(self, params):
+        """obtain transformed thresholds from original thresholds, cutoff
+        constants.
+
+        """
+        start_ppf = params
+        thresh_params = np.concatenate((start_ppf[:1],
+                                        np.log(np.diff(start_ppf[:-1]))))
+        return thresh_params
+
+    def predict(self, params, exog=None):
+        """predicted probabilities for each level of the ordinal endog.
+
+
+        """
+        # structure of params = [beta, constants_or_thresholds]
+
+        # explicit in several steps to avoid bugs
+        th_params = params[-(self.k_levels - 1):]
+        thresh = np.concatenate((th_params[:1],
+                                 np.exp(th_params[1:]))).cumsum()
+        thresh = np.concatenate(([-np.inf], thresh, [np.inf]))
+        xb = self.exog.dot(params[:-(self.k_levels - 1)])[:, None]
+        low = thresh[:-1] - xb
+        upp = thresh[1:] - xb
```

Comment:
maybe not here
predict computes all integration bounds, loglike only two per observation

in predict low up could be selected after computing `thresh - xb` instead of computing it twice